File: add_exit_confirmation.js
const fs = require('fs');
const path = require('path');

const dir = path.join(__dirname, 'crm_app/lib/features/orders');
const files = fs.readdirSync(dir).filter(f => f.endsWith('_form_screen.dart'));

const willPopCode = `
  Future<bool> _onWillPop() async {
    final shouldPop = await showDialog<bool>(
      context: context,
      builder: (context) {
        return AlertDialog(
          title: Text('Are You Sure To Exit ?', style: GoogleFonts.inter(fontWeight: FontWeight.bold)),
          content: Text('Any unsaved progress will be lost.', style: GoogleFonts.inter()),
          actions: [
            TextButton(
              onPressed: () => Navigator.of(context).pop(false),
              child: Text('No', style: GoogleFonts.inter(color: AppTheme.corporateBlue)),
            ),
            ElevatedButton(
              onPressed: () => Navigator.of(context).pop(true),
              style: ElevatedButton.styleFrom(backgroundColor: Colors.red),
              child: Text('Yes', style: GoogleFonts.inter(color: Colors.white)),
            ),
          ],
        );
      },
    );
    return shouldPop ?? false;
  }

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: _onWillPop,
      child: Scaffold(
`;

for (const file of files) {
  const filePath = path.join(dir, file);
  let content = fs.readFileSync(filePath, 'utf8');

  // Check if already applied
  if (content.includes('WillPopScope')) {
    console.log(`Skipped (already has WillPopScope): ${file}`);
    continue;
  }

  // Find build method
  const buildRegex = /@override\s+Widget build\(BuildContext context\)\s*\{\s*return\s+Scaffold\(/m;
  if (!buildRegex.test(content)) {
    console.log(`Could not find standard Scaffold build method in: ${file}`);
    continue;
  }

  content = content.replace(buildRegex, willPopCode.trim() + '\n');
  
  // Need to add closing parenthesis for WillPopScope at the very end of the build method
  // Actually, wait, just wrapping Scaffold means we need to add a closing parenthesis `)` at the end of the Scaffold.
  // The Scaffold in build method usually ends right before the next method or the end of the class.
  // To avoid parsing dart accurately, we can replace the closing brace of the build method.
  // But wait, what if we use a regex to find the end of the Scaffold?
  // Let's look at how the build method ends in these files.
}

File: add_exit_confirmation.py
import os
import re

DIR = r"c:\projects\we_crm\crm_app\lib\features\orders"

will_pop_method = """
  Future<bool> _onWillPop() async {
    final shouldPop = await showDialog<bool>(
      context: context,
      builder: (context) {
        return AlertDialog(
          title: const Text('Are You Sure To Exit ?', style: TextStyle(fontWeight: FontWeight.bold)),
          content: const Text('Any unsaved progress will be lost.'),
          actions: [
            TextButton(
              onPressed: () => Navigator.of(context).pop(false),
              child: const Text('No'),
            ),
            ElevatedButton(
              onPressed: () => Navigator.of(context).pop(true),
              style: ElevatedButton.styleFrom(backgroundColor: Colors.red),
              child: const Text('Yes', style: TextStyle(color: Colors.white)),
            ),
          ],
        );
      },
    );
    return shouldPop ?? false;
  }

"""

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    if '_onWillPop' in content:
        print(f"Skipped {os.path.basename(filepath)}")
        return

    # Find the build method
    build_match = re.search(r'Widget\s+build\s*\(\s*BuildContext\s+context\s*\)\s*\{', content)
    if not build_match:
        return

    # Insert _onWillPop just before build
    build_idx = build_match.start()
    content = content[:build_idx] + will_pop_method + content[build_idx:]

    # Find return Scaffold(
    # Because we added will_pop_method, the index of build changed.
    # We find it again
    build_match = re.search(r'Widget\s+build\s*\(\s*BuildContext\s+context\s*\)\s*\{', content)
    
    # We search for return Scaffold( AFTER the build method
    scaffold_match = re.search(r'return\s+Scaffold\s*\(', content[build_match.start():])
    if not scaffold_match:
        print(f"Could not find return Scaffold in {os.path.basename(filepath)}")
        return
        
    start_scaffold = build_match.start() + scaffold_match.start()
    
    # Find the matching closing parenthesis for Scaffold
    # The string starts at return Scaffold( -> we want the index of (
    open_paren_idx = content.find('(', start_scaffold)
    
    count = 0
    end_scaffold = -1
    for i in range(open_paren_idx, len(content)):
        if content[i] == '(':
            count += 1
        elif content[i] == ')':
            count -= 1
            if count == 0:
                end_scaffold = i
                break
                
    if end_scaffold == -1:
        print(f"Mismatched parentheses in {os.path.basename(filepath)}")
        return
        
    # Replace return Scaffold( with return WillPopScope(onWillPop: _onWillPop, child: Scaffold(
    # And add ) at end_scaffold
    
    new_content = (
        content[:start_scaffold] + 
        "return WillPopScope(\n      onWillPop: _onWillPop,\n      child: Scaffold(" + 
        content[open_paren_idx+1:end_scaffold] + 
        ")" + 
        content[end_scaffold:]
    )
    
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(new_content)
    print(f"Updated {os.path.basename(filepath)}")

for filename in os.listdir(DIR):
    if filename.endswith('_form_screen.dart'):
        process_file(os.path.join(DIR, filename))
